src/libraries/System.Linq/src/System/Linq/Average.cs

```diff
@@ -36,6 +37,32 @@ public static double Average(this IEnumerable<int> source)
             }
         }
 
+        private static double Average(ReadOnlySpan<int> span)
+        {
+            if (span.IsEmpty)
+            {
+                ThrowHelper.ThrowNoElementsException();
+            }
+
+            long sum = 0;
+            int i = 0;
+
+            if (Vector.IsHardwareAccelerated)
+            {
+                for (; i <= span.Length - Vector<int>.Count; i += Vector<int>.Count)
+                {
+                    sum += Vector.Sum(new Vector<int>(span.Slice(i)));
```

Comment:
The reduction of the vector (i.e. `Vector.Sum`) could be done once after this loop.
So just accumulate in the loop, then do the reduction to the scalar value.

Pseudo:
```c#
Vector<int> accumulator = default;
for (...)
{
    accumulator += new Vector<int>(span.Slice(i)));
}
sum = Vector.Sum(accumulator);
```

Reply:
> So just accumulate in the loop, then do the reduction to the scalar value.

It needs to accumulate as longs, not ints.  That would require widening and then adding the resulting long vectors.  Is that going to be measurably better? (I can of course try it.)

Reply:
Arm64 has `AdvSimd.AddWideningLower` and `AdvSimd.AddWideningUpper` which will be extremely efficient here.

x86 would require distinct widen and addition steps, but I expect it will still perform better as vectorized.

Reply:
Thanks.  Using a Widen and then two adds appears to be a bit better, so I changed it to that.